### DISCUSSION TITLE
First stable version and the future of the library

### DISCUSSION THREAD

**Discussion by romm** [CuyZ/Valinor#265] [Announcements]
This topic aims to centralize discussions about the release of the first stable version (check the release notes [here](https://github.com/CuyZ/Valinor/releases/tag/1.0.0)).

---

If you have ideas about missing features, please do not hesitate to comment, I'll be glad to read and answer.

**Comment by Slamdunk** [comment]
Hi, as far as I can tell the only big missing piece of the current Valinor is a `Collaborator`.
Maintaining such a valuable FOSS all by yourself is a path that only leads to madness, despair and grief :stuck_out_tongue:

**Reply by romm** [reply]
Haha, thank you for putting so much trust in the project!

> only leads to madness, despair and grief

I hope I'll be able to outlive all of that. 😁 

On a more serious note, I agree and would be very pleased if someone wanted to learn the internals of the library to be able to co-maintain it.

For now I know that I am able to free enough time for it, whenever I start to be more busy I'll probably try to find someone.

**Comment by shulard** [comment]
First, congrats for the stable release ! This is an awesome work that you've done :wink:.

**Reply by romm** [reply]
🤗

**Comment by TimWolla** [comment]
> If you have ideas about missing features

I was thinking about allowing registration of additional “string types” that would desugar into plain PHP strings after validation, somewhat similarly to custom constructors.

Possible use cases include:

- Decoding binary payloads embedded in JSON using b64 or hex encoding.
- Trimming whitespace from a string (e.g. when the data source is the query string, e.g. $_GET).

Example of a workaround that technically more-or-less works, but isn't _great_, because `B64String` is not actually a string.

```php
<?php declare(strict_types=1);

use CuyZ\Valinor\Mapper\Source\Source;
use CuyZ\Valinor\MapperBuilder;

require('vendor/autoload.php');

final class B64String {
    private string $s;

    public function __construct(string $s) {
        $this->s = base64_decode($s);
    }

    public function __toString()
    {
        return $this->s;
    }
}

final class Target {
    public function __construct(
        public B64String $b64,
        public string $hex,
        public string $trim,
    ) {
    }
}

$rawJson = <<<'EOT'
{
    "b64": "dGVzdA==",
    "hex": "74657374",
    "trim": " test "
}
EOT;

$result = (new MapperBuilder())
    ->mapper()
    ->map(
        Target::class,
        Source::json($rawJson)
    );

var_dump(strtoupper($result->b64));
```

**Reply by romm** [reply]
Hi @TimWolla, better late than never: attribute converters were introduced in #673, allowing the following:

```php
<?php
declare(strict_types=1);

use CuyZ\Valinor\Mapper\AsConverter;
use CuyZ\Valinor\Mapper\Source\Source;
use CuyZ\Valinor\MapperBuilder;

require('vendor/autoload.php');

#[Attribute(Attribute::TARGET_PROPERTY), AsConverter]
final class Base64Decode
{
    public function map(string $value): string
    {
        return base64_decode($value);
    }
}

final class Target
{
    public function __construct(
        #[Base64Decode]
        public string $b64,
        public string $hex,
        public string $trim,
    ) {}
}

$rawJson = <<<'EOT'
{
    "b64": "dGVzdA==",
    "hex": "74657374",
    "trim": " test "
}
EOT;

$result = (new MapperBuilder())
    ->mapper()
    ->map(
        Target::class,
        Source::json($rawJson)
    );

var_dump(strtoupper($result->b64)); // "TEST"
```
